### DISCUSSION TITLE
"Privacy feature". What is the right approach to hide hosted domains linked to mailserver?

### DISCUSSION THREAD

**Discussion by Jarrodsz** [docker-mailserver/docker-mailserver#3312] [Q&A]
After reading the docs I could not really found a solution so far so I hope someone has expierience with this:

Would it be possible to hide a list of all domain names that are hosted on the mailserver?

I knew there are several online "research websites" that list all the domains hosted on an IP block or a specific IP.
That way its possible for someone to retrieve a list of all domain.ltd's that are hosted on a specific mailserver.

If I where to setup my mail DNS like

```mail.mydomain1.com -> ip of mailserver```
or 
```mail.mydomain2.com -< ip of mailserver```

It would be potentially be possible to retrieve the IP of the mailserver and get a list of all all the domains that are hosted on this mailserver ip. 

What would be the best way to prevent this? Would I use another IP for example on a digitalocean Droplet VPS that would point to the same server ( you can add extra IPS to a digitalocean VPS easily) so then I would do:

```mail.mydomain1.com -> ip of mailserver```
or 
```mail.mydomain2.com -< extra ip attached to digitalocean droplet/vps that points to the same VPS as the "ip of mailserver" above```

Would this be a solution to prevent leaking my domains?

This way all the 

```mail.mydomain1.com -> ip of mailserver```

would be possible to "connectable" to 1 single mailserver but only the domains setup in DNS like:

```mail.mydomain2.com -< extra ip attached to digitalocean droplet/vps that points to the same VPS as the "ip of mailserver" above```

would be "connectable" to each other.

Perhaps there is even a better solution to this.
The reason for asking this is due to security and privacy reasons I want to disable an easy way for people to figure out all my domains if they know my mailserver (ip/ips)

**Comment by polarathene** [comment]
A mail server is not tied to the mail domains it manages AFAIK, although it's common to have both under the same domain when you're only managing a single domain.

- `mail.example.com` might not have accounts for `@example.com`, but it could for `@domain.net` and `@whatever.xyz`.
- Those domains have their own DNS with records that direct mail to `mail.example.com` and authorize it to send mail on their behalf. Their IPs in this case are unrelated to the mail servers.

When that is not the case because your mail server shares the same domain for email accounts you want to manage, that may be the same IP, but it doesn't have to be, that part depends on DNS records.

Then you have TLS certificates, these can be looked up in search engines like our TLS docs mention, showing a history of all the domains/subdomains you've provisioned certs for which can help an attacker identify what services you're running if they're self-descriptive (_`mail.example.com`, `grafana.example.com`, `nextcloud.example.com`, etc_), but they could also just do port scans, query DNS, attempt random logins or URL requests (_like for PHP pages, even if it is a web server but doesn't actually serve PHP pages_).

So similar to the common domain of `example.com` listing all certs provisioned (including historically) via services like `crt.sh`, if the IP lookup concerns you sure, mitigate that via a separate IP. But if there is a search based on the IP/domain of the mail-server, and that can likewise return results of DNS records referencing it from other domains, then you're not likely to gain much benefit?

They're only going to be aware of the domains though, not accounts.

---

If you're concerned security wise about that information to assist in login attempts, you'd be better off with enabling service like Fail2Ban to make online login attempts ineffective and reduced load on your server if that becomes an issue.

For securing your accounts, passwords should have a minimum amount of entropy that makes guessing impractical. [getapassphrase.com](https://getapassphrase.com) can help you there. For example here's a password with 44 bits of entropy: `starved pilot buys falling flower`, that means if the attacker knew exactly the rules for generating this password, then their best attack is still a random guess of `2^44` possible passwords (1 out of 17 trillion). DMS (or Dovecot rather) presently uses SHA512-CRYPT to store a hashed output of an account password, this isn't the greatest choice but it's adequate.

If you're interested in the math for such an attack, [I share a fair amount of insight here](https://github.com/FusionAuth/fusionauth-issues/issues/85#issuecomment-1493371525). And you can see [this benchmark for `sha512crypt` on an Nvidia RTX 4080 GPU](https://gist.github.com/bigpick/cfa22947c884f7a3fc1431475e345427), it cites **700K hash rate**. That equates to about **145 days as an offline attack** (`((2^44 / 700e3) / (60*60*24)) / 2`), online there will be much more latency and you'd have Fail2Ban blocking the attackers IP, since the attacker won't have the hashed password to compare to for performing an offline attack. In this scenario 44-bit entropy should be sufficient, but if you'd like to be extra secure go up to something like 64-bit (_the link I provided demonstrates how much of a difference that makes_), or leverage a password manager if you prefer.

**Reply by Jarrodsz** [reply]
I
Inderstand my mistake make total sense noe.

Thank you for this awesome in depth explanation. Im no security newbie but nevertheless its very nice to read up on this as a good reminder and learnt some new things.

Im indeed more concerned that domains can be linked together than mailboxes.

If I understand correctly thus I would be able to have basically only 1 certificate and 1 mailserver domain. Mail.mymailserver.ltd for example which I would use to connect my mail clients on.

I would then have to investigate if i can attach another certificate for mail.mymailsever2.ltd which i could attach to another ip thst links to the same VPS that hosts the mailserver. I then be able to hide" associations beteeen those mailsevers since they would be on another domain.

I would have 2 groups of domains that i setup the dns for. Group 1 would be marketing websites that might be linked to each other. And  another group of domains for business that might be linked to each other. That way it would not be easy to link the marketing and business "group" of domains to each other.

Did i inderstand it correctly like that? 🙏

**Reply by polarathene** [reply]
> Did i inderstand it correctly like that?

You have one mail server, then you only have `mail.example.com`, does not matter how many domains it will manage.

Your other domains DNS then can add MX record like `example.com`, `domain.net`, `whatever.xyz`, and all of them point to `mail.example.com` for delivery. Then they also need other DNS records like TXT record for SPF to say that `mail.example.com` is also authorized to send mail on their behalf.

Only certificate that matters is the `mail.example.com` one for the mail server which needs to prove the server you connected to really is for `mail.example.com`, so the connecting client can trust them. Secure TLS connection does not matter alone if we don't have the trust from certificate. Once we have that, the other domains that are managed have no connections from the client or mail server, so only the DNS records are relevant for more information / trust.

---

Just keep in mind that the search engine you're concerned about, there is probably such for indexing IP and DNS records. So if one exists that lets you search `mail.example.com`, it may know of this for `example.com` DNS, but also `domain.net` and `whatever.xyz` with MX DNS record. I don't think you can really prevent that, once that information is scraped from somewhere or found by automation/scanning.

**Comment by Jarrodsz** [comment]
Still don't get it fully yet.

What I basically would like to have is this:

1. mail.mymailserver.ltd
2. mail.mymailserver2.ltd

Both would link to another IP that connects to the same server ( DigitalOcean Allows for this )

Then,

I could setup my domains to use any of those mailservers in the DNS.

That way I can connect to both mail.mymailserver.ltd and mail.mymailserver2.ltd and get the mail ( both would have a letsencrypt certificate )

Is this possible?

I know I can use 2 docker-mailservers and configure them but it would be beneficial to memory usage and maintenance to have only 1 mailserver that supports this.

**Reply by polarathene** [reply]
> Still don't get it fully yet.
> 
> What I basically would like to have is this:
> 
> 1. mail.mymailserver.ltd
> 2. mail.mymailserver2.ltd

What do you think these two are for? How do you think they work?

Your server with DMS will have a DNS A record for `mail.example.com`, and any domains you have `mymailserver.ltd` + `mymailserver2.tld` can point their MX records to `mail.example.com`. When someone sends mail to `user@mymailserver.ltd` it will lookup the MX record in DNS and be told it lives at `mail.example.com`, then it will query the IP for that server.

Meanwhile you are asking about create your own `mail.domain-here.ltd` MX record for each domain, with their own A record to point to IP of the mail server? (or a different IP shared on the same server)

I don't see the benefit.

---

> That way I can connect to both mail.mymailserver.ltd and mail.mymailserver2.ltd and get the mail ( both would have a letsencrypt certificate )

Just so we are clear. You do not need a certificate for each mail domain you manage with DMS, you only need one for DMS itself.

Like it is a web server, but for mail. AFAIK, DMS can only identify itself as one FQDN (the hostname / origin), this will be an A record in DNS and any domains it manages will have an MX record that should resolve to it.

When something goes wrong like mail delivery failure, DMS will try notify your postmaster account, and it will send it from the FQDN it identifies itself as. When it contacts other mail servers to deliver mail to, it will greet them with this same FQDN, and it will come from one of the public IP address on the server, which will probably be checked against DNS for a match to the DNS A record that DMS said points to it.

I think that last one might be an issue, unless you ensure it always sends mail out on the same IP it has a DNS record for. Any extra IP or DNS record you make like you're suggesting isn't relevant to this interaction.

---

> Is this possible?

I think with extra configuration it might work. I don't think it helps much.

Your concern was about services indexing information about domains, DNS, etc so that it could be searched to link information together that you'd prefer was private.

Since both IP would share the same DMS instance, and the HELO greeting would be the same for both, this would be searchable information if it's indexed.

[Shodan is a popular search service and does check if a mail server is running](https://www.shodan.io/search?query=smtp) - indexing this information, such that it could be searched and identify information. So that provides the common IP addresses on the same server, and another service like [DNSlytics](https://dnslytics.com/reverse-ip) can look those IP up to identify the domains associated to them for example (_or port scan the IP for other services to get more information to reference_).

**Comment by casperklein** [comment]
You could try using a SAN certificate (containing both domain names).

**Reply by polarathene** [reply]
I think their concern was to avoid being able to derive the domains being associated to the same service / machine.

They figured separate IPs and MX records pointing to the same server running DMS would work, but since indexers will query exposed services like SMTP, they'll index the response, and from there the IPs can be searched via shodan or similar services for DNS records / domains that have been indexed for those IP addresses.

I don't think they're going to have much luck with that.

**Comment by Jarrodsz** [comment]
So the only real solutions is running 2 mailservers on different servers shame :( hoped i could run them on 1
